### ISSUE TITLE
Add clipnorm parameter to optimizers

### ISSUE THREAD

**Issue by carlosgmartin** [pytorch/pytorch#56843]
## 🚀 Feature

For each [optimizer](https://pytorch.org/docs/stable/optim.html), add a `clipnorm: float` parameter to step with a clipped gradient norm (default being `inf` for no clipping).

## Alternatives

PyTorch has a [`torch.nn.utils.clip_grad_norm_`](https://pytorch.org/docs/stable/generated/torch.nn.utils.clip_grad_norm_.html) function, but it has disadvantages:

1. It modifies the gradient in-place. Consider a situation where multiple optimizers work on overlapping sets of parameters.

    From a theoretical/correctness standpoint, the clipping should be internal to the optimizer itself, and only the parameter(s) should be modified, rather than their gradients.

2. It is more convenient to set all optimization hyperparameters in the optimizer's constructor. `clip_grad_norm_` requires a manual application before every `.step()`.

3. It requires doing something verbose of the form
    ```python3
    torch.nn.utils.clip_grad_norm_((
        param
        for param_group in optimizer.param_groups
        for param in param_group['params']
    ), 1)
    ```
    in the case where we only have access to `optimizer` itself.

## Additional context

See the `global_clipnorm` keyword argument for [`tf.keras.optimizers.Optimizer`](https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/Optimizer) for Tensorflow's counterpart.
